### ISSUE TITLE
Gen2: DynamoDB Streams - CloudFormation fails due to Circular dependency between resources

### ISSUE THREAD

**Issue by jgo80** [aws-amplify/amplify-category-api#2554]
### Before opening, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-js/issues?q=is%3Aissue+) and [discussions](https://github.com/aws-amplify/amplify-js/discussions).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### JavaScript Framework

Not applicable

### Amplify APIs

GraphQL API

### Amplify Version

v6

### Amplify Categories

auth, function, api

### Backend

Amplify Gen 2 (Preview)

### Environment information

<details>

```
# Put output below this line
  System:
    OS: macOS 14.4.1
    CPU: (10) arm64 Apple M1 Max
    Memory: 4.63 GB / 32.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 22.1.0 - /opt/homebrew/bin/node
    Yarn: 1.22.22 - /opt/homebrew/bin/yarn
    npm: 10.7.0 - /opt/homebrew/bin/npm
    Watchman: 2024.05.06.00 - /opt/homebrew/bin/watchman
  Browsers:
    Chrome: 124.0.6367.119
    Safari: 17.4.1
  npmPackages:
    %name%:  0.1.0 
    @aws-amplify/backend: ^1.0.0 => 1.0.1 
    @aws-amplify/backend-cli: ^1.0.1 => 1.0.2 
    @aws-amplify/react-native: ^1.1.0 => 1.1.0 
    @aws-amplify/ui-react-native: ^2.1.6 => 2.2.0 
    @aws-lambda-powertools/logger: ^2.1.0 => 2.1.0 
    @babel/core: ^7.20.0 => 7.24.5 
    @expo/vector-icons: ^14.0.0 => 14.0.1 
    @react-native-async-storage/async-storage: ^1.23.1 => 1.23.1 
    @react-native-community/netinfo: ^11.3.1 => 11.3.1 
    @react-navigation/native: ^6.0.2 => 6.1.17 
    @types/aws-lambda: ^8.10.137 => 8.10.137 
    @types/jest: ^29.5.12 => 29.5.12 
    @types/react: ~18.2.45 => 18.2.79 (18.3.1)
    @types/react-test-renderer: ^18.0.7 => 18.3.0 
    HelloWorld:  0.0.1 
    aws-amplify: ^6.3.0 => 6.3.0 
    aws-amplify/adapter-core:  undefined ()
    aws-amplify/analytics:  undefined ()
    aws-amplify/analytics/kinesis:  undefined ()
    aws-amplify/analytics/kinesis-firehose:  undefined ()
    aws-amplify/analytics/personalize:  undefined ()
    aws-amplify/analytics/pinpoint:  undefined ()
    aws-amplify/api:  undefined ()
    aws-amplify/api/server:  undefined ()
    aws-amplify/auth:  undefined ()
    aws-amplify/auth/cognito:  undefined ()
    aws-amplify/auth/cognito/server:  undefined ()
    aws-amplify/auth/enable-oauth-listener:  undefined ()
    aws-amplify/auth/server:  undefined ()
    aws-amplify/data:  undefined ()
    aws-amplify/data/server:  undefined ()
    aws-amplify/datastore:  undefined ()
    aws-amplify/in-app-messaging:  undefined ()
    aws-amplify/in-app-messaging/pinpoint:  undefined ()
    aws-amplify/push-notifications:  undefined ()
    aws-amplify/push-notifications/pinpoint:  undefined ()
    aws-amplify/storage:  undefined ()
    aws-amplify/storage/s3:  undefined ()
    aws-amplify/storage/s3/server:  undefined ()
    aws-amplify/storage/server:  undefined ()
    aws-amplify/utils:  undefined ()
    aws-cdk: ^2.140.0 => 2.141.0 
    aws-cdk-lib: ^2.140.0 => 2.141.0 
    constructs: ^10.3.0 => 10.3.0 
    esbuild: ^0.21.1 => 0.21.1 (0.20.2)
    expo: ~51.0.0 => 51.0.2 
    expo-constants: ~16.0.1 => 16.0.1 
    expo-font: ~12.0.4 => 12.0.4 
    expo-linking: ~6.3.1 => 6.3.1 
    expo-router: ~3.5.10 => 3.5.11 
    expo-splash-screen: ~0.27.4 => 0.27.4 
    expo-status-bar: ~1.12.1 => 1.12.1 
    expo-system-ui: ~3.0.4 => 3.0.4 
    expo-web-browser: ~13.0.3 => 13.0.3 
    jest: ^29.2.1 => 29.7.0 
    jest-expo: ~51.0.1 => 51.0.1 
    react: 18.2.0 => 18.2.0 
    react-dom: 18.2.0 => 18.2.0 
    react-native: 0.74.1 => 0.74.1 
    react-native-gesture-handler: ~2.16.1 => 2.16.2 
    react-native-get-random-values: ^1.11.0 => 1.11.0 
    react-native-reanimated: 3.10.0 => 3.10.0 
    react-native-safe-area-context: ^4.10.1 => 4.10.1 
    react-native-screens: 3.31.1 => 3.31.1 
    react-native-url-polyfill: ^2.0.0 => 2.0.0 
    react-native-web: ~0.19.10 => 0.19.11 
    react-test-renderer: 18.2.0 => 18.2.0 
    tsx: ^4.9.3 => 4.9.3 
    typescript: ^5.4.5 => 5.4.5 (4.4.4, 4.9.5)
  npmGlobalPackages:
    eas-cli: 7.4.0
    npm: 10.7.0
    yarn: 1.22.22


```

</details>

### Describe the bug

Follwing the [official Gen 2 Guide](https://docs.amplify.aws/react/build-a-backend/functions/examples/dynamo-db-stream/) to configure a Lambda function with an Amazon DynamoDB stream as an event source I get the following error during deployment/sandbox:

```
The CloudFormation deployment has failed.
Caused By: ❌ Deployment failed: Error [ValidationError]: Circular dependency between resources: [auth179371D7, data7552DF31, function1351588B]


Resolution: Find more information in the CloudFormation AWS Console for this stack.
```

The error happens when I add the Event Source according to the official Docs:
```ts
import { defineBackend } from '@aws-amplify/backend';
import { StartingPosition } from 'aws-cdk-lib/aws-lambda';
import { DynamoEventSource } from 'aws-cdk-lib/aws-lambda-event-sources';
import { auth } from './auth/resource';
import { data } from './data/resource';
import { myDynamoDBFunction } from './functions/dynamoDB-function/resource';

const backend = defineBackend({
  auth,
  data,
  myDynamoDBFunction,
});

const eventSource = new DynamoEventSource(
  backend.data.resources.tables['Todo'],
  {
    startingPosition: StartingPosition.LATEST,
  },
);

backend.myDynamoDBFunction.resources.lambda.addEventSource(eventSource); // <-- this causes the Circular dependency

```
**Edit:** I was concerned my Auth preTokenGeneration handler caused this, but it dit not. When I remove preTokenGeneration from triggers, I still get the error:

```
The CloudFormation deployment has failed.
Caused By: ❌ Deployment failed: Error [ValidationError]: Circular dependency between resources: [data7552DF31, function1351588B]
```

So the circular dependency definitely results from the fact that the backend resource table is referenced in the eventSource and then being injected back into the backend object. Design flaw?

```
const eventSource = new DynamoEventSource(
  backend.data.resources.tables['Todo'], // <- reference from backend
  {
    startingPosition: StartingPosition.LATEST,
  },
);

backend.myDynamoDBFunction.resources.lambda.addEventSource(eventSource); // <- inject back into backend
```

### Expected behavior

It should be possible to add the event source for the lambda according to the docs to invoke the lambda by DynamoDB Stream.

### Reproduction steps

Just follow your own official guide: https://docs.amplify.aws/react/build-a-backend/functions/examples/dynamo-db-stream/

### Code Snippet

```javascript
// Put your code below this line.

```

### Log output

<details>

```
// Put your logs below this line


```

</details>

**Comment by MattWlodarski** [comment]
Hey guys, I get the same error:  

The CloudFormation deployment has failed.
Caused By: ❌ Deployment failed: Error [ValidationError]: Circular dependency between resources: [auth179371D7, data7552DF31, function1351588B]

Here is the code I added to backend.ts: 
```
const SMSResponseEventSource = new DynamoEventSource(backend.data.resources.tables['SMSResponse'], {
  startingPosition: StartingPosition.LATEST,
});
const EmailResponseEventSource = new DynamoEventSource(
  backend.data.resources.tables['EmailResponse'],
  {
    startingPosition: StartingPosition.LATEST,
  },
);
const VoiceCallResponseEventSource = new DynamoEventSource(
  backend.data.resources.tables['VoiceCallResponse'],
  {
    startingPosition: StartingPosition.LATEST,
  },
);

backend.OnAppointmentReminderDeliveryHandler.resources.lambda.addEventSource(
  SMSResponseEventSource,
);
backend.OnAppointmentReminderDeliveryHandler.resources.lambda.addEventSource(
  EmailResponseEventSource,
);
backend.OnAppointmentReminderDeliveryHandler.resources.lambda.addEventSource(
  VoiceCallResponseEventSource,
);
```

I'm hoping this gets fixed soon so I don't have to set this all up manually.

**Comment by MattWlodarski** [comment]
I also get the same error when trying to set up environment variables. The weird things with this is that the error won't always happen immediately. Instead, I'll start getting the error after making other unrelated changes to my sandbox. Then I have to comment out the environment variable code in my backend.ts file to get rid of the circular dependency.

The CloudFormation deployment has failed.
Caused By: ❌ Deployment failed: Error [ValidationError]: Circular dependency between resources: [auth179371D7, data7552DF31, function1351588B]

Commenting out these lines in backend.ts resolves the circular dependency: 

```
const copyDynamoDBTableLambdaFunction = backend.CopyDynamoDBTableHandler.resources
  .lambda as Function;
copyDynamoDBTableLambdaFunction.addEnvironment('APP_SYNC_API_ID', backend.data.apiId);
```

**Comment by palpatim** [comment]
Assigning to @AnilMaktala to try & repro -- I'm not able to get this error to occur.

**Comment by LukaASoban** [comment]
Just wanted to pop in to say that I also am getting the same error when trying to add in event source for dynamodb streams. Same issue circular dependency between auth, data, and function which doesn't make any sense to me since my lambda needs the table but my table doesn't need my lambda?

**Comment by jgo80** [comment]
The issue should be marked as bug rather than as question.

**Comment by LukaASoban** [comment]
@jgo80 do you have a workaround by any chance for now?

**Comment by jgo80** [comment]
> @jgo80 do you have a workaround by any chance for now?
> 
> 

I don't 😞 Just waiting for an answer as a desperate customer moving from Gen1 to Gen2...

**Comment by MattWlodarski** [comment]
I found a similar issue with a resolution: https://github.com/aws-amplify/amplify-backend/issues/1552

I'm not sure if it exactly applies to our situation because I didn't quite follow the recommended solution. Let me know if it helps you guys.

**Comment by LukaASoban** [comment]
Thanks @MattWlodarski I will take a look at this today and give it a try and then update this thread

**Comment by ggj0418** [comment]
Yeah the code provided in the document operates normally when only event mapping is used, but a circular dependency error occurs as soon as data custom query is used.
So if you are using a custom query (both query and mutation), you have to manually make a separate cdk stack for the streaming target table and set the event mapping and policy on that stack.

As @MattWlodarski mentioned, i wrote down that issue and found the solution. the well-worked repository is on my repository (you can find it on that mentioned issue), so you might be able to refer it.

**Comment by LukaASoban** [comment]
I had some triggers in the auth resource. If I remove them, it removes the circular dependency. 

```
triggers: {
    customMessage,
    postConfirmation,
    preSignUp,
    defineAuthChallenge,
    createAuthChallenge,
    verifyAuthChallengeResponse: verifyAuthChallenge,
  },
```

**Comment by LukaASoban** [comment]
@ggj0418 @MattWlodarski do you either of you have any auth triggers?

**Comment by LukaASoban** [comment]
```javascript
import { defineAuth, secret } from "@aws-amplify/backend";
import { customMessage } from "./custom-message/resource";
import { postConfirmation } from "./post-confirmation/resource";
import { preSignUp } from "./pre-signup/resource";
import { defineAuthChallenge } from "./define-auth-challenge/resource";
import { createAuthChallenge } from "./create-auth-challenge/resource";
import { verifyAuthChallenge } from "./verify-auth-challenge/resource";

/**
 * Define and configure your auth resource
 * @see https://docs.amplify.aws/gen2/build-a-backend/auth
 */
export const auth = defineAuth({
  loginWith: {
    email: true,
    phone: true,
  },
  triggers: {
    customMessage,
    postConfirmation,
    preSignUp,
    defineAuthChallenge,
    createAuthChallenge,
    verifyAuthChallengeResponse: verifyAuthChallenge,
  },
});

```

**Comment by ggj0418** [comment]
@LukaASoban hm... no i do not have any auth trigger. did you add that trigger with custom query or something else component of amplify gen2 backend?

**Comment by ggj0418** [comment]
@LukaASoban on another project, im using that auth trigger with data custom queries, but it does not have a circular error

**Comment by LukaASoban** [comment]
```javascript
const backend = defineBackend({
  auth,
  //preSignUp,
  //storage,
  data,
  //getUsersByEmailOrPhoneFunction,
  //createAuthChallenge,
  //checkIfCognitoUserExistsFunction,
  //onUpload,
  dynamodbStreamPlaces,
});

const eventSource = new DynamoEventSource(
  backend.data.resources.tables["Place"],
  {
    startingPosition: StartingPosition.LATEST,
  }
);

backend.dynamodbStreamPlaces.resources.lambda.addEventSource(eventSource);
```

and then my auth resource.ts

```javascript
export const auth = defineAuth({
  loginWith: {
    email: true,
    phone: true,
    // externalProviders: {
    //   signInWithApple: {
    //     clientId: secret("SIWA_CLIENT_ID"),
    //     keyId: secret("SIWA_KEY_ID"),
    //     privateKey: secret("SIWA_PRIVATE_KEY"),
    //     teamId: secret("SIWA_TEAM_ID"),
    //     scopes: ["email"],
    //     attributeMapping: {
    //       email: "email",
    //       custom: {
    //         email_verified: "email_verified",
    //       },
    //     },
    //   },
    //   google: {
    //     clientId: secret("GOOGLE_CLIENT_ID"),
    //     clientSecret: secret("GOOGLE_CLIENT_SECRET"),
    //     scopes: ["email"],
    //     attributeMapping: {
    //       email: "email",
    //       custom: {
    //         email_verified: "email_verified",
    //       },
    //     },
    //   },
    //   callbackUrls: ["http://localhost:5173/", "flaire://"],
    //   logoutUrls: ["http://localhost:5173/", "flaire://"],
    // },
  },
  //accountRecovery: "EMAIL_AND_PHONE_WITHOUT_MFA",
  triggers: {
    customMessage,
    // postConfirmation,
    // preSignUp,
    // defineAuthChallenge,
    // createAuthChallenge,
    // verifyAuthChallengeResponse: verifyAuthChallenge,
  },
});
```

and then in my customMessage trigger function, there is no custom query or dependency on any other component. If I comment out the customMesssage function here it works.

**Comment by ggj0418** [comment]
@LukaASoban I think that dynamoDB event mapping is the point. here is my suggestion.

```typescript
const backend = defineBackend({
  auth,
  //preSignUp,
  //storage,
  data,
  //getUsersByEmailOrPhoneFunction,
  //createAuthChallenge,
  //checkIfCognitoUserExistsFunction,
  //onUpload,
  dynamodbStreamPlaces,
});

const placeTable = backend.data.resources.tables["Place"]
backend.dynamodbStreamPlaces.resources.lambda.role?.attachInlinePolicy(
    new Policy(
        Stack.of(placeTable),
        "DynamoDBPolicy",
        {
            statements: [
                new PolicyStatement({
                    effect: Effect.ALLOW,
                    actions: [
                        "dynamodb:DescribeStream",
                        "dynamodb:GetRecords",
                        "dynamodb:GetShardIterator",
                        "dynamodb:ListStreams",
                    ],
                    resources: ["*"],
                }),
            ],
        }
    )
);

new EventSourceMapping(
    Stack.of(placeTable),
    "test",
    {
        target: backend.dynamodbStreamPlaces.resources.lambda,
        eventSourceArn: placeTable.tableStreamArn,
        startingPosition: StartingPosition.LATEST,
    }
);
```

**Comment by ggj0418** [comment]
Oh i forgot to tell the resolved code's branch. here is the [branch](https://github.com/ggj0418/amplify-next-pages-template/tree/error/circular-dependency). it is not a main. i'm sorry for the missing description.

**Comment by LukaASoban** [comment]
@ggj0418 Thanks for your input here. I seem to be getting this error when I try that:

 CREATE_FAILED        | AWS::Lambda::EventSourceMapping | data/amplifyData/Place/DynamodbStreamPlacesEventSourceMapping (DynamodbStreamPlacesEventSourceMappingE2426243) Resource handler returned message: "Invalid request provided: Stream arn:aws:dynamodb:us-east-2:*********:table/Place-2xgy7oir4fa5tndcygcxeusazu-NONE/stream/2024-05-30T20:36:57.557 is Disabled. You cannot create a lambda mapping on a stream that is Disabled. (Service: Lambda, Status Code: 400, Request ID: 1578ca62-c2ba-4039-ab59-15f48b13ba0e)" (RequestToken: ea51563f-a008-1cb5-0938-f9869001ab6f, HandlerErrorCode: InvalidRequest)

Not sure why my lambda stream is disabled

**Comment by ggj0418** [comment]
@LukaASoban hm... could i see your dynamdbStreamPlaces' handler?

**Comment by LukaASoban** [comment]
Looks exactly like in the docs:

```javascript
import type { DynamoDBStreamHandler } from "aws-lambda";
import { Logger } from "@aws-lambda-powertools/logger";

const logger = new Logger({
  logLevel: "INFO",
  serviceName: "dynamodb-stream-handler",
});

export const handler: DynamoDBStreamHandler = async (event) => {
  for (const record of event.Records) {
    logger.info(`Processing record: ${record.eventID}`);
    logger.info(`Event Type: ${record.eventName}`);

    if (record.eventName === "INSERT") {
      // business logic to process new records
      logger.info(`New Image: ${JSON.stringify(record.dynamodb?.NewImage)}`);
    }
  }
  logger.info(`Successfully processed ${event.Records.length} records.`);
  
  return {
    batchItemFailures: [],
  };
};
```

**Comment by ggj0418** [comment]
@LukaASoban oh yeah sandbox's hotswap is not perfectly worked...
do you talk about the amplify env vars?

**Comment by heclon** [comment]
How come the official docs  publishes a procedure that clearly causes this issue? 
Does no one checks what is being published?

This should be marked as an issue.

**Comment by LukaASoban** [comment]
@heclon are you also getting this same error?

**Comment by MattWlodarski** [comment]
Hey @LukaASoban I also had an auth trigger as well. I appreciate you guys explaining the workaround but I agree with @heclon. This should be marked as an issue and fixed asap. I can't believe they haven't caught this already. They must just test their code in the simplest possible ways. It doesn't take much to trigger either of these circular dependencies...

**Comment by AnilMaktala** [comment]
HI @jgo80, We are able to replicate this issue and marking this as bug for the team to evaluate further.
<img width="1359" alt="image" src="https://github.com/aws-amplify/amplify-category-api/assets/3206171/94201fe4-8268-42a3-93c1-db081044e035">

**Comment by thomazchamberlain** [comment]
I get the exact same error as everyone else here from just following the official docs. Please fix your broken code.

**Comment by ggj0418** [comment]
@leonanderson88 Does the same error occur even if you comment that ses part?

**Comment by LukaASoban** [comment]
@leonanderson88 The below works - here is my code to get this to work to break the circular dependency (you can reduce the IAM role scopes as needed I just gave it all permissions to see if it works)

You basically need to create a custom resource that is a lambda that runs, uses the AWS SDK to get the streamARN for the table and then returns it. You can modify it to include as many tables as you want.

_backend.ts_
```javascript
const userProfileTable = backend.data.resources.tables["UserProfile"];

const dynamodbStreamingLambda =
  backend.dbToOpensearchStreamingFunction.resources.lambda;
dynamodbStreamingLambda.addToRolePolicy(
  new iam.PolicyStatement({
    sid: "AllowStreaming",
    actions: ["dynamodb:*"],
    resources: ["*"],
  }),
);

// getStreamARN Lambda IAM Policy
const getStreamArnsLambdaPolicy = new iam.PolicyStatement({
    actions: ["dynamodb:*"],
    resources: ["*"],
  });


// Create a stack for the custom resources
const customResourceStack = backend.createStack("CustomResourceStack");

// Create a custom resource to init the event mappings (we need to get the stream ARNs)
const getStreamArnsLambda = new lambda.Function(
  customResourceStack,
  "GetStreamArnsLambda",
  {
    runtime: lambda.Runtime.NODEJS_LATEST,
    code: lambda.Code.fromAsset(
      "./amplify/custom/trigger-functions/get-dynamodb-stream-arns",
    ),
    handler: "getStreamArns.handler",
    timeout: Duration.seconds(30),
    initialPolicy: [getStreamArnsLambdaPolicy],
  },
);

// Create a custom resource to init the event mappings (we need to get the stream ARNs for each table)
const streamArnsProvider = new customResources.Provider(
  customResourceStack,
  "StreamArnsProvider",
  {
    onEventHandler: getStreamArnsLambda,
  },
);
const streamArns = new CustomResource(customResourceStack, "StreamArns", {
  serviceToken: streamArnsProvider.serviceToken,
  properties: {
    TableNames: [
      userProfileTable.tableName,
    ], // Add more tables here as we need
  },
});

// Ensure the custom resource depends on the DynamoDB tables creation
streamArns.node.addDependency(userProfileTable);


// Event source mapping for the User Profile table
const userProfileStreamingEventSourceMapping = new lambda.EventSourceMapping(
  customResourceStack,
  "UserProfileStreamingEventSourceMapping",
  {
    target: backend.dbToOpensearchStreamingFunction.resources.lambda,
    eventSourceArn: streamArns.getAttString("UserProfile"),
    startingPosition: StartingPosition.LATEST,
  },
);
userProfileStreamingEventSourceMapping.node.addDependency(streamArns);
```

and then in my getStreamArns lambda

```javascript
/* GetStreamArns lambda function */

const {
  DynamoDBClient,
  DescribeTableCommand,
} = require("@aws-sdk/client-dynamodb");

const client = new DynamoDBClient({ region: process.env.AWS_REGION });

exports.handler = async (event) => {
  const tableNames = event.ResourceProperties.TableNames;

  try {
    const streamArns = {};
    for (const tableName of tableNames) {
      const command = new DescribeTableCommand({ TableName: tableName });
      const data = await client.send(command);
      const baseTableName = tableName.split("-")[0]; // Extract the base name
      streamArns[baseTableName] = data.Table.LatestStreamArn || "";
    }

    console.log("Stream ARNs:", JSON.stringify(streamArns)); // Log the returned ARNs

    return {
      PhysicalResourceId: tableNames.join(","),
      Data: streamArns,
    };
  } catch (error) {
    console.error("Error:", error);

    return {
      PhysicalResourceId: tableNames.join(","),
      Data: {
        Error: error.message,
      },
    };
  }
};
```

**Comment by naedx** [comment]
I've run into the circular dependency issue several times as well in a fairly simple Gen2 project. In the latest case my setup is as follows:



### Setup

```

const backend = defineBackend({
  auth,
  data,
  storage
});

```

I've used `data` category to define the AppSync API from schema. This creates the only AppSync API with the tables (no resolvers). I've defined a function to create all resolvers manually by using `backend.data.addResolver(...)`. I have one resolver which calls a lambda function which needs to access **Storage** to create a presigned URL. So far so good, I'm at:

**Dependency: AppSync >> Storage**

Next I used the [LambdaToSqsToLambda L3 Construct to implement](https://constructs.dev/packages/@aws-solutions-constructs/aws-lambda-sqs-lambda/v/2.61.0?lang=typescript):
1. Storage needs to trigger a Lambda Function to send a message to the SQS queue of on file creation. **Dependency: Storage >> [L3Construct: Lambda Function (producer) >> SQS]**
2. SQS queue needs to notify a Lambda Function (consumer), do processing on the object in Storage, then send the updates to AppsSync. **Dependency: [L3Construct: SQS >> Lambda Function] >> Storage & AppSync**

<details>

```typescript

// Definition of LambdaSQSLambda L3 Construct
// See: https://constructs.dev/packages/@aws-solutions-constructs/aws-lambda-sqs-lambda/v/2.61.0?lang=typescript
//
//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

const lambdaSQSLambdaStack = backend.createStack('LambdaSQSLambdaStack');

const lambdaToSQSToLambdaConstruct = new LambdaToSqsToLambda(lambdaSQSLambdaStack, 'LambdaToSqsToLambdaPattern', {
  producerLambdaFunctionProps: {           // << -- this way doesn't work because it creates the lambda in the same stack
                                           //         as the SQS queue
    runtime: LambdaRuntime.NODEJS_20_X,
    handler: 'index.handler',
    code: fromAssetHelper(path.resolve('amplify/data/features/my-feature/upload/s3-to-sqs-handler.lambda.cjs'))
  },

  consumerLambdaFunctionProps: {           // << -- this way doesn't work because it creates the lambda in the same stack
                                           //         as the SQS queue
    runtime: LambdaRuntime.NODEJS_20_X,
    handler: 'index.handler',
    code: fromAssetHelper(path.resolve('amplify/data/features/my-feature/upload/sqs-to-appsync-handler.lambda.cjs'))
  },

});

// PRODUCER gets notified of uploads and writes the event to the SQS queue
backend.storage.resources.bucket.addObjectCreatedNotification(new LambdaDestination(lambdaToSQSToLambdaConstruct.producerLambdaFunction));

lambdaToSQSToLambdaConstruct.sqsQueue.grantSendMessages(lambdaToSQSToLambdaConstruct.producerLambdaFunction);

// CONSUMER listens to the SQS queue, does processing and sends response to AppSync (data)
backend.data.resources.graphqlApi.grantMutation(
  lambdaToSQSToLambdaConstruct.consumerLambdaFunction,
  '*'
);

backend.storage.resources.bucket.grantRead(
  lambdaToSQSToLambdaConstruct.consumerLambdaFunction,
  '*'
);

lambdaToSQSToLambdaConstruct.consumerLambdaFunction.addToRolePolicy(new PolicyStatement({
  actions: [
    "appsync:GraphQL",
  ],
  resources: [
    `${backend.data.resources.graphqlApi.arn}/*`,
  ],
}));

lambdaToSQSToLambdaConstruct.consumerLambdaFunction.addEnvironment('API_ENDPOINT', backend.data.graphqlUrl);




```

</details>


### Problem

So the circular dependency comes because now I have:

❌ **Dependency: AppSync >> Storage >> [L3Construct: SQS >> Lambda Function] >> Storage & AppSync**


### Solution 

The solution I found was to control the stack that the resources were created in, placing related resources in a common stack. The L3 construct above allows defining and providing a reference to lambda functions. This allows us to reference resources in a different stack. Concretely, this allowed me to define the _producer_ lambda function (which sends data from Storage to SQS) in the `Stack.of(backend.storage.resources.bucket)` and the _consumer_ lambda function (which receives the data from SQS and then uses Storage and AppSync in the `Stack.of(backend.data)`. This leaves me with:

✅ **Dependency: [L3Construct: SQS] >> [Stack: AppSync, Consumer Function] >> [Stack: Storage, Producer Function]**

✅ **Dependency: [L3Construct: SQS] >> [Stack: Storage, Producer Function]**

<details>

```typescript

const lambdaSQSLambdaStack = backend.createStack('LambdaSQSLambdaStack');

const lambdaToSQSToLambdaConstruct = new LambdaToSqsToLambda(lambdaSQSLambdaStack, 'LambdaToSqsToLambdaPattern', {
  // producerLambdaFunctionProps: {           // << -- this way doesn't work because it creates the lambda in the same stack
                                              //       as the SQS queue
  //   runtime: LambdaRuntime.NODEJS_20_X,
  //   handler: 'index.handler',
  //   code: fromAssetHelper(path.resolve('amplify/data/features/my-feature/upload/s3-to-sqs-handler.lambda.cjs'))
  // },
  existingProducerLambdaObj: new LambdaFunction(Stack.of(backend.storage.resources.bucket), 'LambdaToSqsProducerLambda', {
    handler: 'index.handler',
    code: fromAssetHelper(path.resolve('amplify/data/features/my-feature/upload/s3-to-sqs-handler.lambda.cjs')),
    runtime: LambdaRuntime.NODEJS_20_X,
  }),
  // consumerLambdaFunctionProps: {           // << -- this way doesn't work because it creates the lambda in the same stack
                                              //       as the SQS queue
  //   runtime: LambdaRuntime.NODEJS_20_X,
  //   handler: 'index.handler',
  //   code: fromAssetHelper(path.resolve('amplify/data/features/my-feature/upload/sqs-to-appsync-handler.lambda.cjs'))
  // },
  existingConsumerLambdaObj: new LambdaFunction(Stack.of(backend.data), 'LambdaToSqsConsumerLambda', {
    handler: 'index.handler',
    code: fromAssetHelper(path.resolve('amplify/data/features/my-feature/upload/sqs-to-appsync-handler.lambda.cjs')),
    runtime: LambdaRuntime.NODEJS_20_X,
    // environment: {
    // }
  })
});

```

</details>


This broke the dependency for me.

**Comment by naedx** [comment]
TLDR: as a workaround you could try defining your lambda function manually (instead of defineFunction()) and placing it in `Stack.of(data)` where your tables are. That way, there aren't an inter stack dependencies.

**Comment by MyNameIsTakenOMG** [comment]
I also bumped into this issue when trying to grant dynamodb table permissions to my lambda handler or pass table name into lambda as an env variable. And for me, it is quite difficult to break this circular reference, so I'm thinking about giving custom resource a try just as what @LukaASoban shared( thank you for sharing this workaround!)

Personally, I think this is a bug that has a big impact when working with gen2. Hopefully it could be fixed soon

**Comment by naedx** [comment]
@MyNameIsTakenOMG , I also found that ampx doesn't automatically take care of our .ts files when not created using `defineFunction()`. 

I've created [a helper function](https://github.com/naedx/amplify-playground/tree/dev/projects/copy-lambda-module-helper) to copy the .cjs file that is tranpiled from my .ts file to the correct asset folder that ampx has designated for my function. It then gets packaged and uploaded. It could be modified to include the transpile step directly.

**Comment by MyNameIsTakenOMG** [comment]
> @MyNameIsTakenOMG , I also found that ampx doesn't automatically take care of our .ts files when not created using `defineFunction()`.
> 
> I've created [a helper function](https://github.com/naedx/amplify-playground/tree/dev/projects/copy-lambda-module-helper) to copy the .cjs file that is tranpiled from my .ts file to the correct asset folder that ampx has designated for my function. It then gets packaged and uploaded. It could be modified to include the transpile step directly.

Oh, many thanks 🙏  I was looking for a way to transpile the **ts** code and put it into the correct place so that `ampx` could pick up and upload. I will definitely check it out. Thank you, really appreciate it 👍

**Comment by codemachine24** [comment]
Is this going to be fixed? Or is this the solution? https://github.com/ggj0418/amplify-next-pages-template/compare/main...error/circular-dependency

Trying understand where this issue is, because it's a blocker for what we're trying to solve.

**Comment by Orf1** [comment]
bump, my team and I aren't able to progress a production application due to the issue!

**Comment by godrimcom** [comment]
Same issue here. I'm using a custom mutation (not a function)
```
deleteTickets: a
    .mutation()
    .arguments({
      ids: a.id().array(),
    })
    .returns(a.json())
    .authorization((allow) => [allow.authenticated()])
    .handler(
      a.handler.custom({
        dataSource: a.ref('Ticket'),
        entry: './handlers/tickets/deleteTickets.js',
      })
    ),
```

**Comment by godrimcom** [comment]
> bump, my team and I aren't able to progress a production application due to the issue!

me too

**Comment by ggj0418** [comment]
I think it is not a good idea to use `scheduled Job` or `data stream` feature with `amplify gen2`.
Our team are in a situation where `scheduled jobs` are essential, so we had no choice but to use `aws sam & aws cdk` to implement the feature. I hope this function will be officially supported by `amplify gen2` ASAP.

**Comment by eliasericsson** [comment]
I submit to you a way to reproduce the error from the provided starter template:
1. Follow instructions for "DynamoDB Streams" ✅ [example](https://github.com/eliasericsson/circular-dependency-bug/commit/d320756a04bf11c5e681309caec13a5c9b018476?diff=unified&w=1) 
2. Follow instructions for "Add user to group" ❌ [example](https://github.com/eliasericsson/circular-dependency-bug/commit/2a4f5540e236237b72c4e87b8385687cee8d5458?diff=unified&w=1)

```shell
The CloudFormation deployment has failed.
Caused By: ❌ Deployment failed: Error [ValidationError]: Circular dependency between resources: [auth179371D7, data7552DF31, function1351588B]


Resolution: Find more information in the CloudFormation AWS Console for this stack.
```

PS. The resolution part of the message above is misleading, there is no more information in the CloudFormation console. DS.

**Comment by dpilch** [comment]
The docs have been updated. https://docs.amplify.aws/react/build-a-backend/functions/examples/dynamo-db-stream/

The root cause is that all functions are placed in a single CFN stack. There will be a circular dependency between the data and function stack if:
1. There is a dependency from data to any functions in the app.
2. There is dependency from any function to any data resource.

I'll be updating https://github.com/aws-amplify/amplify-category-api/issues/2375 soon with a recommendation on how to move a function to a different stack.
